docs/design/features/source-generator-pinvokes.md

```diff
@@ -0,0 +1,233 @@
+
+In this system it is not defined how marshaling of specific types would be performed. The built-in runtime has complex rules for some types, and it is these rules that once shipped become the de facto standard - often times regardless if the behavior is a bug or not. The design here is not concerned with how the arguments go from a managed to unmanaged environment.
+
+### Source Generator constraint mitigations
+
+In the current Source Generator design modification of any user written code is not permitted. This includes modification of any non-functional metadata (e.g. Attributes). The above design mentions modification of user source so the current design may be untenable. Let us define the following options and describe how the above design would be modified.
+
+* In-memory source modification - This would yield the ideal design above.
+
+* On disk source modification - Altering source on disk would require the `GenerateNativeImportAttribute` to contain all properties from `DllImportAttribute`. This is required since removing the `DllImportAttribute` would be impossible and thus would cause issues during build since the method must also be marked as `partial`.
+
+* No source modification - This aligns with the current Source Generator constraints. It would be possible to create a [Roslyn Analyzer and Code fix](https://github.com/dotnet/roslyn/wiki/Getting-Started-Writing-a-Custom-Analyzer-&-Code-Fix) to aid the developer in converting their `DllImportAttribute` marked functions to use `GenerateNativeImportAttribute`. Furthermore, the function would need to be updated to have the `partial` keyword and potentially the enclosing class.
+
+## Proposed APIs
+
+The ideal solution with support for minor user defined source updates using Source Generators:
+
+``` CSharp
+namespace System.Runtime.InteropServices
+{
+    /// <summary>
+    /// Attribute used to indicate a Source Generator should create a function for marshaling
+    /// arguments instead of relying on the CLR to generate an IL Stub at runtime.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = false, Inherited = false)]
+    public sealed class GenerateNativeImportAttribute : Attribute
+    {
+    }
+}
+```
+
+An alternative solution that merges the existing `DllImportAttribute` with the `GenerateNativeImportAttribute`. This option would be used if user source cannot be modified in any manner.
+
+``` CSharp
+namespace System.Runtime.InteropServices
+{
+    /// <summary>
+    /// Attribute used to indicate a Source Generator should create a function for marshaling
+    /// arguments instead of relying on the CLR to generate an IL Stub at runtime.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = false, Inherited = false)]
+    public sealed class GenerateNativeImportAttribute : Attribute
```

Comment:
I think that this name should as similer to `DllImportAttribute` as possible, e.g.  `GeneratedDllImportAttribute`.

Reply:
Sure.